### PR TITLE
Adding synchronization and threading primitives

### LINKED ISSUE #1482: Add tracy mutex annotations
We already want to replace absl::Mutex for code-size reasons, so it may make sense as part of that to also just add the right tracy annotations so that we can get mutex contention information. Section 3.5 of the tracy docs have more information.

### PR DESCRIPTION
**Pull request by benvanik**
These will be used to implement the task system used to execute CPU kernels. The mutex benchmarks show we are faster than std::mutex (due to some limitations we can accept) and we now have complete control over any cross-thread blocking behavior we want to introduce (to favor latency vs. power consumption, etc).

This is compatible with Windows, iOS/MacOS (following the Apple performance/power guidelines), Android/Linux, wasm/emscripten, and fallbacks for normal pthreads usage on other platforms. I've only optimized Windows right now but the other platforms shouldn't require much tweaking and when they do we have a benchmark to do it with.

Progress on #49.
Fixes #1482.

### CODE REVIEW

**Review by ScottTodd — APPROVED** [review]

File: iree/base/atomics.h
Comment (benvanik) [reply iree/base/atomics.h:45]: yep - we may also be able to query these from cpuinfo but I wanted to wait until that landed separately first

**Pull request merged**
